core/src/main/clojure/xtdb/operator/patch.clj

```diff
@@ -1,27 +1,40 @@
 (ns xtdb.operator.patch
   (:require [clojure.spec.alpha :as s]
             [xtdb.error :as err]
+            [xtdb.expression :as expr]
             [xtdb.logical-plan :as lp]
             [xtdb.time :as time]
             [xtdb.types :as types]
-            [xtdb.vector.writer :as vw])
+            [xtdb.util :as util]
+            [xtdb.vector.writer :as vw]
+            )
   (:import org.apache.arrow.vector.types.pojo.Schema
            org.apache.arrow.vector.VectorSchemaRoot
            xtdb.operator.PatchGapsCursor
-           [xtdb.vector RelationReader]))
+           [xtdb.vector RelationReader]
+           (org.apache.arrow.memory BufferAllocator)))
 
 (s/def ::instantable
   (s/or :literal ::time/datetime-value
-        :param ::lp/param))
+        :param ::lp/param
+        :expr ::lp/expression))
 
-(s/def ::valid-from (s/nilable ::instantable))
-(s/def ::valid-to (s/nilable ::instantable))
+(s/def ::valid-from ::instantable)
+(s/def ::valid-to ::instantable)
 
-(defn- ->instant [[tag value] {:keys [^RelationReader args] :as opts}]
+(defn- ->instant [[tag value] {:keys [^RelationReader args ^BufferAllocator allocator default] :as opts}]
   (-> (case tag
         :literal value
         :param (-> (.vectorForOrNull args (str value))
-                   (.getObject 0)))
+                   (.getObject 0))
+        :expr (let [param-types (expr/->param-types args)
+                    projection (expr/->expression-projection-spec "_temporal_expression"
+                                                                  (expr/form->expr value {:param-types param-types})
+                                                                  {:param-types param-types})]
+                (util/with-open [res (.project projection allocator vw/empty-args {} args)]
+                  (let [inst-like (or (.getObject res 0) default)]
+                    (time/expect-instant inst-like)
+                    inst-like))))
```

Comment:
I am adding an expression in the RA as it make planning in `sql.clj` easier, this is currently not exposed in SQL PATCH.